packages/dds/tree/src/changeset/toDelta.ts

```diff
@@ -0,0 +1,362 @@
+                        }
+                        break;
+                    }
+                    case "MoveIn": break;
+                    case "Bounce":
+                    case "Intake":
+                        fail(ERR_NOT_IMPLEMENTED);
+                        break;
+                    default: unreachableCase(type);
+                }
+            }
+        } else {
+            // Inline into `switch(mark.type)` once we upgrade to TS 4.7
+            const type = mark.type;
+            switch (type) {
+                case "Modify": {
+                    if (mark.tomb === undefined) {
+                        out.push({
+                            offset,
+                            mark: {
+                                type: Delta.MarkType.Modify,
+                                ...convertModify<Delta.OuterMark>(mark),
+                            },
+                        });
+                    }
+                    break;
+                }
+                case "Delete": {
+                    if ("count" in mark) {
+                        const deleteMark: Delta.Delete = {
+                            type: Delta.MarkType.Delete,
+                            count: mark.count,
+                        };
+                        out.push({ offset, mark: deleteMark });
+                    } else {
+                        const fields = convertModify<Delta.ModifyDeleted | Delta.MoveOut>(mark).fields;
+                        if (fields !== undefined) {
+                            const deleteMark: Delta.ModifyAndDelete = {
+                                type: Delta.MarkType.ModifyAndDelete,
+                                fields,
+                            };
+                            out.push({ offset, mark: deleteMark });
+                        } else {
+                            const deleteMark: Delta.Delete = {
+                                type: Delta.MarkType.Delete,
+                                count: 1,
+                            };
+                            out.push({ offset, mark: deleteMark });
+                        }
+                    }
+                    break;
+                }
+                case "MoveOut":
+                case "Revive":
+                case "Return":
+                case "Gap":
+                    fail(ERR_NOT_IMPLEMENTED);
+                case "Tomb": {
+                    // These tombs are only used to precisely describe the location of other attaches.
+                    // They have no impact on the current state.
+                    break;
+                }
+                default: unreachableCase(type);
+            }
+        }
+    }
+    // TODO: add runtime checks
+    return out as unknown as Delta.PositionedMarks<TMarks>;
+}
+
+/**
+ * Clones the content described by a Changeset into tree content expected by Delta.
+ */
+function cloneTreeContent(content: ProtoNode[]): Delta.ProtoNode[] {
```

Comment:
We plan to replace ProtoNode with ITreeCursor soon, right?

Reply:
Yes :)